### PR TITLE
BUG: Incorrect variable window bounds for first row when offset cover all rows

### LINKED ISSUE #42753: BUG: Inconsistent result from centered datetime-like rolling window aggregates
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
import pandas as pd

def rolling_test(window, df=None, center=True):
    if df is None:
        df = pd.DataFrame(
            {"x": 1},
            index=[pd.Timestamp("20130101 09:00:01"),
                   pd.Timestamp("20130101 09:00:02")
            ] 
        )
    
    df["right"] = df.rolling(window, closed="right", center=center).x.sum()
    df["both"] = df.rolling(window, closed="both", center=center).x.sum()
    df["left"] = df.rolling(window, closed="left", center=center).x.sum()
    df["neither"] = df.rolling(window, closed="neither", center=center).x.sum()
    
    return df

rolling_test('1T')
```

#### Problem description

When `center=True` and the offset is large enough to contain all of the rows, the first output row of rolling aggregates is wrong most of the time.  This can happen for all possible values of the `closed` parameter.  For the sample code the observed wrong results are `1` when `closed` is `right` or `both` and `NaN` when `closed` is `left` or `neither`.  Here are some examples:

```
                     x  right  both  left  neither
2013-01-01 09:00:01  1    1.0   1.0   NaN      NaN
2013-01-01 09:00:02  1    2.0   2.0   2.0      2.0

                     x  right  both  left  neither
2013-01-01 09:00:01  1    2.0   2.0   2.0      NaN
2013-01-01 09:00:02  1    2.0   2.0   2.0      2.0

                     x  right  both  left  neither
2013-01-01 09:00:01  1    1.0   2.0   2.0      2.0
2013-01-01 09:00:02  1    2.0   2.0   2.0      2.0
```

I have tested with `sum` and `count` aggregates and the problem seems to be that the first row is taken by itself.  The same behavior is present on 1.3.1 and on master (b8d750fb90c68d4af25af89442aba001483adffa).

#### Expected Output

The code sample uses an offset of `1T` and the data only has two seconds worth of data points so any aggregate should include all the rows.  Sometimes I get the correct result:

```
                     x  right  both  left  neither
2013-01-01 09:00:01  1    2.0   2.0   2.0      2.0
2013-01-01 09:00:02  1    2.0   2.0   2.0      2.0
```

Even if I were wrong about the expected result, there is the fact that the result vary from one moment to the next.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : b8d750fb90c68d4af25af89442aba001483adffa
python           : 3.9.4.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.6.0
Version          : Darwin Kernel Version 19.6.0: Thu May  6 00:48:39 PDT 2021; root:xnu-6153.141.33~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.4.0.dev0+305.gb8d750fb90
numpy            : 1.21.1
pytz             : 2021.1
dateutil         : 2.8.2
pip              : 21.0.1
setuptools       : 54.2.0
Cython           : 0.29.24
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by dicristina**
- [x] closes #42753 
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

---

Allow for cases in which the end bound of the variable window for the first row extends past the last row, that is, cases in which the offset is large enough to cover all the rows.

### CODE REVIEW

File: pandas/_libs/window/indexers.pyx [hunk 1]
```diff
@@ -79,11 +79,11 @@ def calculate_variable_window_bounds(
     else:
         end[0] = 0
     if center:
-        for j in range(0, num_values + 1):
-            if (index[j] == index[0] + index_growth_sign * window_size / 2 and
-               right_closed):
+        for j in range(0, num_values):
+            if ((index[j] < index[0] + index_growth_sign * window_size / 2) or
```
Comment (jreback) [inline pandas/_libs/window/indexers.pyx:83]: can you use a temporary variable for the rhs

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
cc @mroeschke

**Comment by jreback** [comment]
@dicristina can you merge master

File: pandas/tests/window/test_rolling.py [hunk 2]
```diff
@@ -220,6 +220,36 @@ def test_datetimelike_centered_selections(
...
+    index = [
+        Timestamp("20130101 09:00:01"),
+        Timestamp("20130101 09:00:02"),
+        Timestamp("20130101 09:00:02"),
+    ]
+    df = frame_or_series([1, 1, 1], index=index)
+
+    result = df.rolling(window, closed=closed, center=True).sum()
+    expected = frame_or_series(expected, index=index)
+    tm.assert_equal(result, expected, check_dtype=False)
```
Comment (mroeschke) [inline pandas/tests/window/test_rolling.py:250]: Why is `check_dtype=False` here?

Comment (dicristina) [reply pandas/tests/window/test_rolling.py:250]: There is no good reason for that.  Thanks.

File: doc/source/whatsnew/v1.3.2.rst [hunk 3]
```diff
@@ -32,6 +32,7 @@ Bug fixes
 ~~~~~~~~~
 - Bug in :meth:`pandas.read_excel` modifies the dtypes dictionary when reading a file with duplicate columns (:issue:`42462`)
 - 1D slices over extension types turn into N-dimensional slices over ExtensionArrays (:issue:`42430`)
+- Fixed bug in :meth:`Series.rolling` and :meth:`DataFrame.rolling` not calculating window bounds correctly for the first row when window is an offset that covers all the rows (:issue:`42753`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.3.2.rst:35]: Good to mention here specifically for when `center=True`

**Review by mroeschke — APPROVED** [review]
One comment otherwise LGTM

**Review by mroeschke — APPROVED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @dicristina

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x

**Pull request merged**